pkg/util/tlsutil/certgen.go

```diff
@@ -0,0 +1,236 @@
+// Copyright Confidential Containers Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package tlsutil
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"math/big"
+	"time"
+)
+
+// Certificate generation is based on https://github.com/golang/go/blob/master/src/crypto/tls/generate_cert.go
+
+// Automatic mutual TLS configuration works as follows
+//
+// 1. At start up, cloud-api-adaptor generates the following two sets of certificate/key pairs
+//    * Self-signed client certificate and its private key
+//    * Self-signed server CA certificate and its private key
+// 2. Before creating a peer pod VM, cloud-api-adaptor generates a pair of server certificate and private key using the server CA certificate.
+// 3. When creating a peer pod VM, cloud-api-adaptor sends the generated server certificate/key as well as client certificate to a newly created pod VM via cloud-init data
```

Comment:
Shouldn't the pod VM use server certificate/key and ca certificate ?

Reply:
For server authentication, since there are multiple servers (peer pod VMs) per worker node, cloud-api-adaptor generates a single server CA certificate, and then issues a server certificate from it for each peer pod VM.

For client authentication, since there is only a single client (cloud-api-adaptor process) per worker node, I don't think cloud-api-adaptor needs to generate a client CA certificate. Instead, cloud-api-adaptor generates a self-signed client certificate, and agent-protocol-forwarder uses the client certificate as a root certificate for client authentication.

Reply:
Hmm, I was thinking the flow like this. 
1. Create a self-signed root CA (ca.crt)
2. Create client (client.crt, client.key) and server (server.crt, server.key) certificates, signed by root CA
3. CAA sends the server.crt, server.key and ca.crt to the pod VM via cloud-init. 
Instead of  CAA sending the server.crt, server.key and client.crt which is what I understood from your earlier comments.

Reply:
The flow In my current implementation is like this.
1. Create a self-signed root CA (ca.crt, ca.key) **only for server certificates to be issued**
2. Create a self-signed **client** certificate (client.crt, client.key)
3. Issue a server certificate (server.crt, server.key) signed by the root server  CA (ca.key)
4. CAA sends server.crt, server.key (as server-side key) and client.crt (as client root certificate) to the pod VM via cloud-init

In general, we can use difference trusted root certificate pools for server and client authentications in mTLS, although it is completely valid to use the single same root certificate for both client and server authentications.

In my current implementation, I don't use the same certificate as client and server root certificates, just because we don't have to do so. I just use separated trust chains in client and server authentications. Of course, we can use the same root certificate for both server and client authentications, but I don't come up with any strong reason to do so.

If we use the same CA to issue a client certificate, both server and client certificates are validated by the same CA certificate. It means that a peer pod VM can use the server certificate to connect to another peer pod VM by mimicking a client certificate. This is very factitious scenario, and I am not sure what does this actually mean, but separation of trust chain in client and server authentications will prevent this kind of misuse.

@bpradipt do you have any idea regarding the use of the same CA for both client and server authentications?

Reply:
In all the self-signed certificate scenarios that I have seen, I saw single root CA being used to sign all the certificates. Kind of mimicking trusted root CA used within an organisation. Hence the flow of not using the root `ca.crt` in the cert chain felt like an anti-pattern to me. But honestly I'm no expert in this, and what you mentioned above also makes sense. So I'll leave it to your discretion :-)